### ISSUE TITLE
Plugin system

### ISSUE THREAD

**Issue by nodiscc** [shaarli/Shaarli#14]
**update: See https://github.com/shaarli/Shaarli/wiki/Ideas-for-plugins**

Hey, I did a roundup of issues on https://github.com/sebsauvage/Shaarli/issues and tried to list requests/bugs that would fit as plugins. Then see what we have:

### Add/remove page elements
* ~~https://github.com/sebsauvage/Shaarli/issues/51: Make timestamps optional~~ **implemented**
* ~~https://github.com/sebsauvage/Shaarli/issues/113: RSS/Atom: make permalinks optional~~ **FIXED in https://github.com/shaarli/Shaarli/pull/63**
* https://github.com/sebsauvage/Shaarli/issues/100: redirect permalink to the linked page (instead of the shaarli entry) (use shaarli as URL shortener)
* ~~https://github.com/sebsauvage/Shaarli/issues/81: piwik support (custom HTML head)~~ **closed by requester**
* ~~https://github.com/sebsauvage/Shaarli/issues/168: Remove annoying URL parameters (facebook, utm-source...)~~ **fixed in 01b8f52718f5136fb7de54e09b47ae4523aec9e8**
* https://github.com/sebsauvage/Shaarli/issues/11: Add weekly and monthly pages (like the daily page).
* ~~https://github.com/sebsauvage/Shaarli/pull/105: Adds landmarks (geo-tags) to posts _(implemented)_~~ **no reply, please open a new issue if interested**
* ~~https://github.com/sebsauvage/Shaarli/pull/95: Adds a stylesheet selector~~ **moved to #22**
* https://github.com/sebsauvage/Shaarli/issues/154: Add custom page feature
* ~~https://github.com/sebsauvage/Shaarli/issues/59: New bookmarklet for note taking~~ **moved to https://github.com/shaarli/Shaarli/issues/142**
* ~~https://github.com/sebsauvage/Shaarli/issues/7: add a Read later button~~ **moved to https://github.com/shaarli/Shaarli/issues/143**
* ~~https://github.com/sebsauvage/Shaarli/pull/162: add "Play videos" button in toolbar~~ **implemented, waiting in [`new-plugin-system`](https://github.com/shaarli/Shaarli/tree/new-plugin-system) branch**
* ~~https://github.com/sebsauvage/Shaarli/issues/28: There should be only one "feed" button (make buttons optional)~~ **fixed**
* https://github.com/sebsauvage/Shaarli/issues/153: RSS/ATOM: append tags to the feed item description
* ~~https://github.com/sebsauvage/Shaarli/issues/187: pubsubhubbub wrongly supported~~ **fixed**
* ~~https://github.com/sebsauvage/Shaarli/pull/144: Support coloration syntaxique et balise de code _(implémenté, ajoute [Geshi](http://qbnz.com/highlighter/) et un include dans index.php)_~~ **please open a new issue if interested**
* https://github.com/sebsauvage/Shaarli/pull/128: Option for enable HTML in descriptions, including WYSIWYG _(implémenté, ajoute [TinyMCE](http://www.tinymce.com/) et un script dans page.footer.html)_
 * https://github.com/sebsauvage/Shaarli/issues/177: Mise en forme de la description / Customization of description _(duplicate?)_
* ~~https://github.com/sebsauvage/Shaarli/pull/158: Add origin/via field~~ **-> WONTFIX, not in core, maybe as a plugin, documented in wiki**
* ~~https://github.com/sebsauvage/Shaarli/issues/71: Video playback support in picture wall~~ **implemented, , waiting in [`new-plugin-system`](https://github.com/shaarli/Shaarli/tree/new-plugin-system) branch**
* https://github.com/sebsauvage/Shaarli/issues/129: Add get parameters 'description' and 'tags' on new entries creation
* ~~https://github.com/sebsauvage/Shaarli/issues/152: (Tag autocompletion based on description)~~ **corner case, wontfix**
* https://github.com/sebsauvage/Shaarli/issues/83: Tag suggestion based on content of the link
* https://github.com/sebsauvage/Shaarli/issues/157: signaler un Lien mort
* https://github.com/sebsauvage/Shaarli/issues/189: Daily tag cloud
* https://github.com/sebsauvage/Shaarli/issues/184: Tree style display (filterable tag cloud?)
* ~~https://github.com/sebsauvage/Shaarli/issues/75: Pingback support _(partially implemented)_~~ **please open a new issue if interested**
* ~~https://github.com/sebsauvage/Shaarli/issues/203: Display timestamps on daily page~~ **FIXED in 38a2d03e34a309abee2b243738fdce1c4cb0be16**

#### Sharing tools
* ~~https://github.com/sebsauvage/Shaarli/pull/70: Publish to Twitter (with bit.ly) and Facebook~~ **please open a new issue if interested, should be a generic social service sharing plugin, without user tracking**
* ~~https://github.com/sebsauvage/Shaarli/issues/171: Make QR code optional~~ **moved to https://github.com/shaarli/Shaarli/issues/57**
* ~~https://github.com/sebsauvage/Shaarli/issues/14: Use local QR-Code generation instead of external service~~ **fixed**
 

#### Thumbnailers
* ~~https://github.com/sebsauvage/Shaarli/issues/91: Reduce size of big thumbnails~~ **INVALID, server configuration problem, worksforme**
* ~~https://github.com/sebsauvage/Shaarli/issues/127: Miniature Deviantart~~ **FIXED once https://github.com/shaarli/Shaarli/pull/128 is merged**
* ~~https://github.com/sebsauvage/Shaarli/issues/169: thumbnails or favicons~~ **FIXED once https://github.com/shaarli/Shaarli/pull/128 is merged**

#### Themes
* ~~https://github.com/sebsauvage/Shaarli/issues/93: Create a list of some remarkable CSS for shaarli~~ **FIXED in https://github.com/shaarli/Shaarli-themes**
* ~~https://github.com/sebsauvage/Shaarli/issues/82:~~ **no reply, https://github.com/shaarli/shaarli-themes**
* ~~https://github.com/sebsauvage/Shaarli/issues/159: Would like to submit a theme~~ **FIXED at https://github.com/shaarli/shaarli-themes/pull/2**

#### Comment systems

* ~~https://github.com/sebsauvage/Shaarli/issues/170:~~ Ajout d'un système de commentaires **Latest discussion at https://github.com/shaarli/Shaarli/issues/181.**
* ~~https://github.com/sebsauvage/Shaarli/issues/185: Discussion about a link~~ **closed/ghost**

**Comment by e2jk** [comment]
I have just created a page on the wiki to [list plugin suggestions](https://github.com/shaarli/Shaarli/wiki/Ideas-for-plugins), as a way to unclutter the issues list. But that's before I had realized we had this issue open. Should we delete the wiki page instead?
The idea behind the wiki page is to be able to close issues for plugins that nobody is picking up at the moment, to have a somewhat cleaner issues list, which enables to focus on what is broken. First quality, then quantity...

**Comment by nodiscc** [comment]
@e2jk Thanks for creating the wiki page; we can move "wontfix" plugin requests there (eg. https://github.com/shaarli/Shaarli/issues/58). Just make sure the wiki page is linked from somewhere (the wiki Home page, or the README)

> Should we delete the wiki page instead

I'd keep both for the moment, this issue is a great way to see what would be _required_ for a plugin system, as the current proposal https://github.com/shaarli/Shaarli/pull/52 is not complete (I'll have to rebase/redo it on top of the latest changes). We can close this issue when the plugin system is implemented.

For example, judging by the list at the top:

 * I think we need a plugin system as a **way to inject components in** the HTML head (piwik), in the toolbar (configurable custom page link, JS button), in the linklist info bar (links to external sharing/comments/archival services, qr, Link to dead link signalling ...), in the footer (JS includes).
 * We _may_ need to enable adding new fields to the editlink dialog via plugins, to allow them to set custom properties for links (thumbnail URL? "via" field? Custom comment url? geo-URI ?) via plugins, but it's not as clear. The `savedb()` function would also need to be changed to allow extra fields (currently they are hardcoded).
 * Some ideas listed at the top would probably be better implemented in Shaarli's core (new bookmarklets, tagcloud improvments), or marked `wontfix` because they would require complex plugin infrastructure or are insecure (HTML descriptions support, syntax highlighting...)

**Comment by nodiscc** [comment]
Plugin system pull request at https://github.com/shaarli/Shaarli/pull/129

**Comment by nodiscc** [comment]
I've closed #129. The plugin system is available in branch https://github.com/shaarli/Shaarli/tree/new-plugin-system of this repository. Feel free to add your work on top of it and send Pull Requests. Items that need to be dealt with before merging:

 * QR code onClick event is broken
 * plugin option defaults should not be defined in index.php, rather, explain they have to be configured (in the wiki)bsp@TE:~/git/Shaarli/tpl/plugins (new-plugin-system %=)$

I'll sort more items from the original list and forward people here.

**Comment by nodiscc** [comment]
Clean Pull Request back at https://github.com/shaarli/Shaarli/pull/164

**Comment by nodiscc** [comment]
Blocked by https://github.com/shaarli/Shaarli/issues/163

**Comment by ArthurHoaro** [comment]
> Blocked by #163

Why? Also, #163 if 1.0 milestone. We can work on it later.

**Comment by ArthurHoaro** [comment]
Closing this with #275.

Please open separate issues for plugin suggestion.
